File: token/economy/burn.go
package economy

import (
	"math"

	"github.com/shopspring/decimal"
)

// ==================== 燃烧配置 ====================

// BurnConfig 燃烧配置
type BurnConfig struct {
	FeeBurnPercent       float64       // 交易费燃烧百分比
	RewardBurnThreshold  decimal.Decimal // 奖励燃烧阈值
	RewardBurnPercent    float64       // 奖励燃烧百分比
	GovernanceBurnPercent float64      // 治理燃烧百分比
}

func DefaultBurnConfig() *BurnConfig {
	threshold, _ := decimal.NewFromString("10000000000000000000") // 10000 POLE
	return &BurnConfig{
		FeeBurnPercent:       0.25,  // 25%
		RewardBurnThreshold:  threshold,
		RewardBurnPercent:    0.10,  // 10%
		GovernanceBurnPercent: 0.01,  // 1%
	}
}

// ==================== 代币燃烧 ====================

// TokenBurn 代币燃烧
type TokenBurn struct {
	config      *BurnConfig
	totalBurned decimal.Decimal
}

// NewTokenBurn 创建燃烧器
func NewTokenBurn(config *BurnConfig) *TokenBurn {
	if config == nil {
		config = DefaultBurnConfig()
	}
	
	return &TokenBurn{
		config:      config,
		totalBurned: decimal.Zero,
	}
}

// CalculateFeeBurn 计算交易费燃烧量
func (tb *TokenBurn) CalculateFeeBurn(fee decimal.Decimal) decimal.Decimal {
	return fee.Mul(decimal.NewFromFloat(tb.config.FeeBurnPercent))
}

// CalculateRewardBurn 计算奖励燃烧量
func (tb *TokenBurn) CalculateRewardBurn(reward decimal.Decimal) decimal.Decimal {
	if reward.GreaterThan(tb.config.RewardBurnThreshold) {
		excess := reward.Sub(tb.config.RewardBurnThreshold)
		return excess.Mul(decimal.NewFromFloat(tb.config.RewardBurnPercent))
	}
	return decimal.Zero
}

// CalculateGovernanceBurn 计算治理燃烧量
func (tb *TokenBurn) CalculateGovernanceBurn(locked decimal.Decimal) decimal.Decimal {
	return locked.Mul(decimal.NewFromFloat(tb.config.GovernanceBurnPercent))
}

// Burn 燃烧代币
func (tb *TokenBurn) Burn(amount decimal.Decimal) {
	tb.totalBurned = tb.totalBurned.Add(amount)
}

// GetTotalBurned 获取总燃烧量
func (tb *TokenBurn) GetTotalBurned() decimal.Decimal {
	return tb.totalBurned
}

// ProcessFee 处理交易费
func (tb *TokenBurn) ProcessFee(fee decimal.Decimal) (burnAmount, remaining decimal.Decimal) {
	burnAmount = tb.CalculateFeeBurn(fee)
	remaining = fee.Sub(burnAmount)
	tb.Burn(burnAmount)
	return
}

// ProcessReward 处理奖励
func (tb *TokenBurn) ProcessReward(reward decimal.Decimal) (burnAmount, remaining decimal.Decimal) {
	burnAmount = tb.CalculateRewardBurn(reward)
	remaining = reward.Sub(burnAmount)
	tb.Burn(burnAmount)
	return
}

// ==================== 铸造配置 ====================

// MintConfig 铸造配置
type MintConfig struct {
	InitialInflation float64 // 初始通胀率
	DecayFactor     float64 // 衰减因子
	TargetTNGV      float64 // 目标 TNGV
	BlockTime       uint64  // 区块时间
}

func DefaultMintConfig() *MintConfig {
	return &MintConfig{
		InitialInflation: 0.20,
		DecayFactor:     0.5,
		TargetTNGV:      1000000.0,
		BlockTime:       3,
	}
}

// ==================== 代币铸造 ====================

// TokenMinter 代币铸造
type TokenMinter struct {
	config       *MintConfig
	currentYear uint32
}

// NewTokenMinter 创建铸造器
func NewTokenMinter(config *MintConfig) *TokenMinter {
	if config == nil {
		config = DefaultMintConfig()
	}
	
	return &TokenMinter{
		config:       config,
		currentYear: 1,
	}
}

// GetInflationRate 获取通胀率
func (tm *TokenMinter) GetInflationRate() float64 {
	// Annual_Inflation_Rate = Initial_Inflation × (1/2)^(Year / 2)
	exponent := float64(tm.currentYear) / 2.0
	return tm.config.InitialInflation * math.Pow(tm.config.DecayFactor, exponent)
}

// CalculateAnnualReward 计算年奖励
func (tm *TokenMinter) CalculateAnnualReward(totalSupply decimal.Decimal) decimal.Decimal {
	rate := tm.GetInflationRate()
	return totalSupply.Mul(decimal.NewFromFloat(rate))
}

// CalculateBlockReward 计算区块奖励
func (tm *TokenMinter) CalculateBlockReward(totalSupply decimal.Decimal) decimal.Decimal {
	annual := tm.CalculateAnnualReward(totalSupply)
	blocksPerYear := 31536000 / int(tm.config.BlockTime)
	return annual.Div(decimal.NewFromInt(int64(blocksPerYear)))
}

// CalculateEpochReward 计算纪元奖励
func (tm *TokenMinter) CalculateEpochReward(totalSupply decimal.Decimal, blocksPerEpoch uint64) decimal.Decimal {
	blockReward := tm.CalculateBlockReward(totalSupply)
	return blockReward.Mul(decimal.NewFromInt(int64(blocksPerEpoch)))
}

// AdjustForTNGV 根据 TNGV 调整奖励
func (tm *TokenMinter) AdjustForTNGV(baseReward decimal.Decimal, currentTNGV float64) decimal.Decimal {
	ratio := tm.config.TargetTNGV / max(currentTNGV, 1.0)
	adjustment := math.Sqrt(ratio)
	// 限制在 0.5x 到 2x 之间
	if adjustment < 0.5 {
		adjustment = 0.5
	}
	if adjustment > 2.0 {
		adjustment = 2.0
	}
	
	return baseReward.Mul(decimal.NewFromFloat(adjustment))
}

// AdvanceYear 推进年份
func (tm *TokenMinter) AdvanceYear() {
	tm.currentYear++
}

// GetCurrentYear 获取当前年份
func (tm *TokenMinter) GetCurrentYear() uint32 {
	return tm.currentYear
}

File: token/economy/src/burn.rs
//! Token Burn Module

use crate::types::*;
use std::sync::Arc;
use tokio::sync::RwLock;

/// Token burn mechanism
pub struct TokenBurn {
    /// Total burned tokens
    total_burned: Arc<RwLock<TokenAmount>>,
    /// Burn configuration
    config: BurnConfig,
}

#[derive(Debug, Clone)]
pub struct BurnConfig {
    /// Transaction fee burn percentage
    pub fee_burn_percent: f64,
    /// Reward burn threshold
    pub reward_burn_threshold: TokenAmount,
    /// Reward burn percentage
    pub reward_burn_percent: f64,
    /// Governance burn percentage
    pub governance_burn_percent: f64,
}

impl Default for BurnConfig {
    fn default() -> Self {
        Self {
            fee_burn_percent: 0.25,        // 25%
            reward_burn_threshold: 10_000_000_000_000_000_000u128, // 10,000 POLE
            reward_burn_percent: 0.10,     // 10%
            governance_burn_percent: 0.01,  // 1%
        }
    }
}

impl TokenBurn {
    pub fn new(config: BurnConfig) -> Self {
        Self {
            total_burned: Arc::new(RwLock::new(0)),
            config,
        }
    }

    /// Calculate fee burn amount
    pub fn calculate_fee_burn(&self, fee: TokenAmount) -> TokenAmount {
        (fee as f64 * self.config.fee_burn_percent) as TokenAmount
    }

    /// Calculate reward burn amount
    pub fn calculate_reward_burn(&self, reward: TokenAmount) -> TokenAmount {
        if reward > self.config.reward_burn_threshold {
            let excess = reward - self.config.reward_burn_threshold;
            (excess as f64 * self.config.reward_burn_percent) as TokenAmount
        } else {
            0
        }
    }

    /// Calculate governance burn amount
    pub fn calculate_governance_burn(&self, locked: TokenAmount) -> TokenAmount {
        (locked as f64 * self.config.governance_burn_percent) as TokenAmount
    }

    /// Burn tokens
    pub async fn burn(&self, amount: TokenAmount) {
        let mut total = self.total_burned.write().await;
        *total += amount;
    }

    /// Get total burned
    pub async fn get_total_burned(&self) -> TokenAmount {
        *self.total_burned.read().await
    }

    /// Process transaction fee
    pub async fn process_fee(&self, fee: TokenAmount) -> (TokenAmount, TokenAmount) {
        let burn_amount = self.calculate_fee_burn(fee);
        let remaining = fee - burn_amount;
        
        self.burn(burn_amount).await;
        
        (burn_amount, remaining)
    }

    /// Process reward
    pub async fn process_reward(&self, reward: TokenAmount) -> (TokenAmount, TokenAmount) {
        let burn_amount = self.calculate_reward_burn(reward);
        let remaining = reward - burn_amount;
        
        self.burn(burn_amount).await;
        
        (burn_amount, remaining)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_fee_burn() {
        let burn = TokenBurn::new(BurnConfig::default());
        
        let fee = 1000u128;
        let (burn_amount, remaining) = (
            burn.calculate_fee_burn(fee),
            fee - burn.calculate_fee_burn(fee)
        );
        
        assert_eq!(burn_amount, 250);
        assert_eq!(remaining, 750);
    }

    #[test]
    fn test_reward_burn_below_threshold() {
        let burn = TokenBurn::new(BurnConfig::default());
        
        let reward = 5_000_000_000_000_000_000u128; // 5000 POLE < threshold
        let burn_amount = burn.calculate_reward_burn(reward);
        
        assert_eq!(burn_amount, 0);
    }

    #[test]
    fn test_reward_burn_above_threshold() {
        let burn = TokenBurn::new(BurnConfig::default());
        
        let reward = 15_000_000_000_000_000_000u128; // 15000 POLE > threshold
        let burn_amount = burn.calculate_reward_burn(reward);
        
        // (15000 - 10000) * 10% = 500
        assert_eq!(burn_amount, 500_000_000_000_000_000u128);
    }
}
